mlflow/tracking/client.py

```diff
@@ -1418,8 +1419,200 @@ def _is_plotly_figure(fig):
+            image: The image object to be logged.
+            step: Integer training step (iteration) at which the image was saved.
+                Defaults to 0.
+            timestamp: Time when this image was saved. Defaults to the current system time.
+            artifact_file: Legacy. Specifies the path, in POSIX format, where the image
+                will be stored as an artifact relative to the run's root directory (for
+                example, "dir/image.png"). This parameter is kept for backward compatibility
+                and should not be used together with `key`, `step`, or `timestamp`.
+            *args: Positional arguments placeholder. Please refer to the documentation above.
+            **kwargs: Keyword arguments placeholder. Please refer to the documentation above.
+
+        .. code-block:: python
+            :caption: Time-stepped image logging numpy example
+
+            import mlflow
+            import numpy as np
+
+            image = np.random.randint(0, 256, size=(100, 100, 3), dtype=np.uint8)
+
+            run = client.create_run(experiment_id="0")
+            client.log_image(run.info.run_id, "dogs", image, step=3)
+
+        .. code-block:: python
+            :caption: Time-stepped image logging pillow example
+
+            import mlflow
+            from PIL import Image
+
+            image = Image.new("RGB", (100, 100))
+
+            run = client.create_run(experiment_id="0")
+            client.log_image(run.info.run_id, "dogs", image, step=3)
+
+        .. code-block:: python
+            :caption: Legacy artifact file image logging numpy example
+
+            import mlflow
+            import numpy as np
+
+            image = np.random.randint(0, 256, size=(100, 100, 3), dtype=np.uint8)
+
+            run = client.create_run(experiment_id="0")
+            client.log_image(run.info.run_id, image, "image.png")
+
+        .. code-block:: python
+            :caption: Legacy artifact file image logging pillow example
+
+            import mlflow
+            from PIL import Image
+
+            image = Image.new("RGB", (100, 100))
+
+            run = client.create_run(experiment_id="0")
+            client.log_image(run.info.run_id, image, "image.png")
+
+        Raises:
+            TypeError: If unexpected parameters are provided. Please refer to the function
+            documentation for required parameters.
+        """
+        import numpy as np
+        from PIL.Image import Image
+
+        image_over_steps_params = [
+            ("key", str),
+            ("image", (np.ndarray, Image)),
+            ("step", int),
+            ("timestamp", int),
+        ]
+        image_over_steps_required = ["key", "image"]
+
+        image_as_artifact_params = [
+            ("image", (np.ndarray, Image)),
+            ("artifact_file", str),
+        ]
+        image_as_artifact_required = ["image", "artifact_file"]
+
+        def _check_types(param_types, *args, **kwargs):
```

Comment:
this part is hard to read, let's use explicit arg names, and check them directly.